### PR TITLE
Create a customized Multihead attention model for TP

### PR DESCRIPTION
**Pull request by fduwjj**
During TP work, we found that there are some special logic we need to enable for TP in the multi-head attention module. 

For example, sometimes, we need to implicitly change the sharding dim and DTensor view without any comms.
Also, to ensure mathematically equivalence, we also need to make the local tensor's stride same as the sharded tensor.

This is the first step for upstreaming TP code and work. We also added a unit test for multi-head attention module to compare the unsharded one and sharded one and shows that math-wise, they are equal. 

We use the same signature of torch.nn.MultiheadAttention as defined here in: https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html. Since the full logic is too much for one PR, we just focus on self-attention for now.

### CODE REVIEW

File: test/spmd/tensor/test_megatron_example.py [hunk 1]
```diff
@@ -96,18 +115,57 @@ def aggregate_output(outputs):
...
-    def test_simple_megatron_e2e(self):
+    def test_mlp_megatron_e2e(self):
         LR = 0.5
         inp_size = [5, 10]
         torch.manual_seed(0)
         inp = torch.rand(*inp_size, device=self.device_type)
         torch.manual_seed(5)
-        model = SimpleModel(self.device_type)
+        model = MLPModule(self.device_type)
         torch.manual_seed(5)
```
Comment (anj-s) [inline test/spmd/tensor/test_megatron_example.py:166]: Why do we reset the manual_seed here after setting it in line 164?

File: test/spmd/tensor/test_megatron_example.py [hunk 2]
```diff
@@ -96,18 +115,57 @@ def aggregate_output(outputs):
...
+        distribute_tensor(m.qkv.weight, device_mesh, col_wise_sharding)
+    )
+    m.proj.weight = torch.nn.Parameter(
+        distribute_tensor(m.proj.weight, device_mesh, row_wise_sharding)
+    )
+    if m.qkv.bias is not None:
+        m.qkv.bias = torch.nn.Parameter(
+            distribute_tensor(m.qkv.bias, device_mesh, col_wise_sharding)
+        )
+    if m.proj.bias is not None:
```
Comment (anj-s) [inline test/spmd/tensor/test_megatron_example.py:139]: nit: I would group projection params together for readability

File: test/spmd/tensor/test_megatron_example.py [hunk 3]
```diff
@@ -96,18 +115,57 @@ def aggregate_output(outputs):
...
+        m.proj.bias = torch.nn.Parameter(
+            distribute_tensor(m.proj.bias, device_mesh, replicate)
+        )
+    m = _replicate_input_tensor(m, device_mesh, replicate)
+    m.proj = _aggregate_local_tensor(m.proj)
+    m.qkv.weight.register_hook(functools.partial(_gradient_hook, m.qkv.weight))
```
Comment (anj-s) [inline test/spmd/tensor/test_megatron_example.py:145]: Why do we need to do this for the TP case? I don't think I set a grad hook for any of my examples.

File: test/spmd/tensor/test_megatron_example.py [hunk 4]
```diff
@@ -96,18 +115,57 @@ def aggregate_output(outputs):
...
         torch.manual_seed(0)
         inp = torch.rand(*inp_size, device=self.device_type)
         torch.manual_seed(5)
-        model = SimpleModel(self.device_type)
+        model = MLPModule(self.device_type)
         torch.manual_seed(5)
-        model_tp = SimpleModel(self.device_type)
-        model_tp = shard_module(model_tp, self.device_type)
+        model_tp = MLPModule(self.device_type)
+        shard_mlp(model_tp, self.device_type)
```
Comment (anj-s) [inline test/spmd/tensor/test_megatron_example.py:168]: Initialize the optimizer here before running the forward pass?

File: test/spmd/tensor/test_megatron_example.py [hunk 5]
```diff
@@ -129,6 +187,44 @@ def test_simple_megatron_e2e(self):
         output_tp = model_tp(inp)
```
Comment (anj-s) [inline test/spmd/tensor/test_megatron_example.py:187]: I would just do a 2 step loop to check that we can do a FW + BW post optimizer.step(). That way we can check if the grads match as well as the params for both TP and non-TP

File: test/spmd/tensor/test_megatron_example.py [hunk 6]
```diff
@@ -129,6 +187,44 @@ def test_simple_megatron_e2e(self):
...
         self.assertEqual(output, output_tp)
 
+    @with_comms
+    def test_self_attn_megatron_e2e(self):
+        LR = 0.05
+        inp_size = [8, 12, 16]
+        torch.manual_seed(0)
+        inp = torch.rand(*inp_size, device=self.device_type)
+        torch.manual_seed(5)
+        model = TensorParallelMultiheadAttention(
```
Comment (anj-s) [inline test/spmd/tensor/test_megatron_example.py:197]: Should this be initialized to the non TP MHA version?

File: test/spmd/tensor/test_megatron_example.py [hunk 7]
```diff
@@ -129,6 +187,44 @@ def test_simple_megatron_e2e(self):
...
+
+        output.sum().backward()
+        output_tp.sum().backward()
+
+        optim = torch.optim.SGD(model.parameters(), lr=LR)
+        optim.step()
+        optim = torch.optim.SGD(model_tp.parameters(), lr=LR)
+        optim.step()
+
+        torch.manual_seed(3)
```
Comment (anj-s) [inline test/spmd/tensor/test_megatron_example.py:222]: Same comments as the previous example.

**Comment by anj-s** [comment]
Thank you for the PR! nice work :) Left a few questions and suggestions.

File: test/spmd/tensor/test_megatron_example.py (hunk 1 above)
Comment (fduwjj) [reply test/spmd/tensor/test_megatron_example.py:166]: Because we want to ensure models are initialized the same way.

File: spmd/tensor/parallel/multihead_attention_tp.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    Multi-head Attention block from Transformer models.
+    Since we need some customizations for the attention layer,
+    we are writing a customized but mathematically equivalent
+    attention module as defined in torch.nn.
+
+    Note that:
+    We now only support the case when it's self attention with
+    limited input args and we also assume that the input tensor
+    has a dimension of three. Although we do implement the logic
+    for multihead attention, it was not fully tested.
```
Comment (XilunWu) [inline spmd/tensor/parallel/multihead_attention_tp.py:50]: Do we need to extend this test to cover the case in future?

File: test/spmd/tensor/test_megatron_example.py [hunk 9]
```diff
@@ -29,11 +20,39 @@ def forward(self, x):
...
+    module.register_forward_pre_hook(hook_func)
+    return module
+
+
 def _gradient_hook(param, grad):
     param._local_tensor.grad = grad._local_tensor
 
 
-def shard_module(m, device_type):
+def shard_mlp(m, device_type):
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply test/spmd/tensor/test_megatron_example.py:55]: Agree with @anj-s. If `dim_groups` is absent and `mesh` is 1D, then `DeviceMesh` uses the default process group.

File: test/spmd/tensor/test_megatron_example.py (hunk 1 above)
Comment (XilunWu) [reply test/spmd/tensor/test_megatron_example.py:166]: My understanding is resetting `manual_seed` ensures that `model` and `model_tp` have the same random input.

File: test/spmd/tensor/test_megatron_example.py (hunk 3 above)
Comment (fduwjj) [reply test/spmd/tensor/test_megatron_example.py:145]: This is indeed needed because it's for 2D parallel. The way FSDP was designed that they will use gradients from local tensors and we need to pass gradients from DT to local tensors. cc: @wanchaol

File: test/spmd/tensor/test_megatron_example.py (hunk 4 above)
Comment (fduwjj) [reply test/spmd/tensor/test_megatron_example.py:168]: Out of curiosity, is there any reason for that?

File: test/spmd/tensor/test_megatron_example.py (hunk 5 above)
Comment (fduwjj) [reply test/spmd/tensor/test_megatron_example.py:187]: I think we are already doing that?

File: test/spmd/tensor/test_megatron_example.py (hunk 7 above)
Comment (fduwjj) [reply test/spmd/tensor/test_megatron_example.py:222]: This is creating the input for post-optimization.

File: spmd/tensor/parallel/multihead_attention_tp.py (hunk 8 above)
Comment (fduwjj) [reply spmd/tensor/parallel/multihead_attention_tp.py:50]: Yep, we will add more tests down the road as we implemented more features for Multihead Attention.

File: test/spmd/tensor/test_megatron_example.py (hunk 5 above)
Comment (anj-s) [reply test/spmd/tensor/test_megatron_example.py:187]: I am suggesting the following for the sake of readability:
1. Initialize model
2. Initialize optimizer
3. for each step in the loop
-> check initial params match
-> check grads match
-> check params match post step()

Right now by running a FW post step() call, you are trying to verify the same criteria as the above FW pass.

File: test/spmd/tensor/test_megatron_example.py (hunk 7 above)
Comment (anj-s) [reply test/spmd/tensor/test_megatron_example.py:222]: Yes, I am suggesting we use a training loop closer to how it is usually written.

File: test/spmd/tensor/test_megatron_example.py (hunk 4 above)
Comment (anj-s) [reply test/spmd/tensor/test_megatron_example.py:168]: I only mention it as a way to improve readability. Wrote down what I meant in more detail below.

File: spmd/tensor/_view_n_reshard.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        ctx,
+        self: "spmd_tensor.DTensor",
+        sharding_dim,
+        shape,
+    ):
+        ctx.previous_placement = self.placements
+        ctx.previous_device_mesh = self.device_mesh
+        ctx.previous_local_shape = self.to_local().size()
+        if (
+            self.placements[0].is_shard(dim=sharding_dim)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/_view_n_reshard.py:22]: QQ, do we have any use cases for multi-dim mesh?

File: test/spmd/tensor/test_megatron_example.py [hunk 11]
```diff
@@ -4,23 +4,14 @@
...
-    distribute_tensor,
-    DeviceMesh,
-    DTensor,
-    Shard,
-    Replicate,
-)
-
-
-class SimpleModel(torch.nn.Module):
+from spmd.testing.common_utils import DistTensorTestBase, with_comms  # type: ignore
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/spmd/tensor/test_megatron_example.py:7]: I feel this test indeed belong to tp. Maybe we can move the whole test under the new folder?

File: test/spmd/tensor/test_megatron_example.py (hunk 1 above)
Comment (fduwjj) [reply test/spmd/tensor/test_megatron_example.py:166]: Because we initialize the module twice. We can definitely move the resetting seed into the module initialization.

File: test/spmd/tensor/test_megatron_example.py (hunk 5 above)
Comment (fduwjj) [reply test/spmd/tensor/test_megatron_example.py:187]: I see. Got it.

File: spmd/tensor/_view_n_reshard.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply spmd/tensor/_view_n_reshard.py:22]: Yeah I think we don't in the case of 2-D Parallel if user use it that way, so maybe let's just add a check throwing if it's being passed a device mesh with more than 1d?

File: .github/workflows/spmd_gpu_tests.sh [hunk 12]
```diff
@@ -18,6 +18,7 @@ pip3 --version
 # Turn off progress bar to save logs
 pip3 install --upgrade pip
 pip3 install --pre torch --extra-index-url https://download.pytorch.org/whl/nightly/cu113
+pip3 install hypothesis
```
Comment (fduwjj) [inline .github/workflows/spmd_gpu_tests.sh:21]: Looks like we need to install this for CI to enable test randomness.

File: .github/workflows/spmd_tests.yaml [hunk 13]
```diff
@@ -33,7 +33,7 @@ jobs:
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
-          pip install pytest-shard numpy
+          pip install pytest-shard numpy hypothesis
```
Comment (fduwjj) [inline .github/workflows/spmd_tests.yaml:36]: We need to have this to enable randomness for CPU test.

File: .github/workflows/code-quality.yml [hunk 14]
```diff
@@ -27,6 +27,9 @@ jobs:
         pip install --pre torchvision -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
         pip install git+https://github.com/pbelevich/transformers.git@compatible_with_pt_master
         pip install "black<23" pylint==v3.0.0a5 mypy==v0.902 flake8 pyre-check==0.9.15
+    - name: Pyre check for SPMD
```
Comment (fduwjj) [inline .github/workflows/code-quality.yml:30]: Move Pyre to the first place so that we can get signals faster. Mypy runs about 10 mins.

File: test/spmd/tensor/parallel/test_tp_examples.py [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+                    )
+                    module.register_parameter("bias", replicated_bias)
+                    module.bias.register_hook(
+                        functools.partial(_gradient_hook, module.bias)
+                    )
+
+    def replicate_input(inputs):
+        DTensors = []
+        for tensor in inputs:
+            DTensors.append(DTensor(tensor, device_mesh, replicate))
```
Comment (wanchaol) [inline test/spmd/tensor/parallel/test_tp_examples.py:165]: why we can using `DTensor` constructor here instead of `from_local`? Let's not use DTensor constructor directly in non-op impls, as we probably need to make the __new__ constructor be more restrictive and only be used internally (we need to add things like shape, stride to the constructor as it will be hard to use by caller)

**Review by wanchaol — COMMENTED** [review]
Looks good, have a few more comments

File: spmd/tensor/parallel/_view_with_dim_change.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import math
+
+import torch
+import spmd.tensor.api as spmd_tensor
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/parallel/_view_with_dim_change.py:5]: Oh, let me correct this. I copied the code from old PR for this API.

File: spmd/tensor/parallel/multihead_attention_tp.py [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+def _stride_same_as_shard(
+    tensor: torch.Tensor, tp_size: int, chunk_dim: int, cat_dim: int
+) -> torch.Tensor:
+    """
+    Adjust local tensor's stride same as the sharded situation.
+    So that view result will keeps the same.
+    """
+    if isinstance(tensor, DT):
+        return tensor
+    view_size = list(tensor.size())
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/parallel/multihead_attention_tp.py:24]: This is not equivalent to `nn.MultiheadAttention` indeed and having this is to make non-DT mode equivalent to DT mode.

File: test/spmd/tensor/parallel/test_tp_examples.py (hunk 15 above)
Comment (fduwjj) [reply test/spmd/tensor/parallel/test_tp_examples.py:165]: Sure. Let's use the public API.

File: spmd/tensor/parallel/multihead_attention_tp.py [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+        value: Union[torch.Tensor, DT],
+        key_padding_mask: Optional[torch.Tensor] = None,
+        need_weights: bool = True,
+        attn_mask: Optional[torch.Tensor] = None,
+        average_attn_weights: bool = True,
+    ) -> Union[torch.Tensor, DT]:
+        b, sq, h = query.shape
+        sk = key.size(1)
+        nh = self.num_heads
+        hn = self.hidden_size_per_attention_head
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/parallel/multihead_attention_tp.py:121]: Will add check in the self attention part.

File: spmd/tensor/parallel/multihead_attention_tp.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/parallel/multihead_attention_tp.py:27]: Add a TODO for adding a test to test equivalence.

File: spmd/tensor/parallel/_view_with_dim_change.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+from spmd.tensor.placement_types import Shard
+from typing import Tuple, Union
+
+
+def _view_with_sharding_dim_change(
+    tensor: Union[torch.Tensor, DT], sharding_dim: int, shape: Tuple[int, ...]
+) -> Union[torch.Tensor, DT]:
+    """
+    We change the implicit sharding dim for a distributed tensor without comms.
+    Because if we don't change sharding dim, we will end up having an invalid DTensor.
```
Comment (wanchaol) [inline spmd/tensor/parallel/_view_with_dim_change.py:15]: "Because if we don't change sharding dim, we will ended up having more comms that are not necessary. Note that this op will produce invalid DTensor, you will need to call this op in pair to recover it back to a valid DTensor."

**Review by wanchaol — APPROVED** [review]
lgtm, one suggestion about the comment.
